Feature/#49 node factories with common configs (#62)
Knotx/knotx-fragments: handler/README.md

configuration `taskNameKey` key
- it finds [task](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#task) factory 
from `taskFactories` that can create a task
- task factory constructs a task (DAG - directed acyclic graph)
- [Fragment Engine](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine) continues 
further fragment's processing

Detailed description of each configuration option is described in the next subsection.

### TaskFactoryOptions
The `taskFactories` is an array of [TaskFactoryOptions](https://github.com/Knotx/knotx-fragments/blob/feature/%2349-node-factories-with-common-configs/handler/core/docs/asciidoc/dataobjects.adoc#taskfactoryoptions) 
that contains configs for registered [task factories](https://github.com/Knotx/knotx-fragments/blob/master/handler/core/src/main/java/io/knotx/fragments/task/TaskFactory.java) 
implementations.

All task factory implementations are registered using a simple service-provider loading facility - 
[Service Loader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html).

# Default task factory

> It is the default task factory containing a list of supported task's names with their definition. A 

Reviewer: ```suggestion
It is the default task factory containing a list of supported tasks' names with their definition. A 
```

It introduces its custom graph nodes
- [action](#action-node)
- [subtasks](#subtasks-node)

## How does it work
Task factory creates a task based on its configuration. It registers graph node factories, delegates 
node initialization to them and joins all nodes with transitions. 

The task is an [identifiable graph](https://github.com/Knotx/knotx-fragments/blob/master/handler/engine/src/main/java/io/knotx/fragments/engine/Task.java) 
that describes the way fragment should be processed. It is a part of Fragment Engine's API.

## How to configure
For all configuration fields and their defaults consult [DefaultTaskFactoryConfig](https://github.com/Knotx/knotx-fragments/blob/master/handler/core/docs/asciidoc/dataobjects.adoc#defaulttaskfactoryconfig).

In general:
- it defines supported tasks (by name) containing graph configuration in `tasks`
- a graph configuration is converted to task with node factories defined in `nodeFactories`

### Tasks
[Tasks](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#task) are configured in 
the form of a dictionary (`taskName -> graph logic`):
```hocon
tasks {
  # unique task name
  myTask { 
    # graph configuration
  }
}
```

> A graph configuration starts with a root [node](#nodes) definition. Each node sets a fragment's logic 

Reviewer: ```suggestion
A graph configuration starts with a root [node](#nodes) definition. Each node sets some logic to perform over a fragment 
```

> to perform and outgoing edges (called [transitions](#transition)). So there are two sections:

Reviewer: ```suggestion
 and outgoing edges (called [transitions](#transition)). So there are two sections:
```

```hocon
node {
  # node options (fragment processing logic)
}
onTransitions {
  # node outgoing edges
}
```

#### Node
Node is described [here](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#node).

> Each task provider can provided its custom nodes.

Reviewer: ```suggestion
Each task provider can provide its custom nodes.
```

Read more about the benefits [here](http://knotx.io/blog/configurable-integrations/).

> ## How to configure

Reviewer: Please update docs.

The above example specifies the action node that delegates processing to the `reference-to-action` 
action and has no transitions.

> The default task factory allows simplifying action nodes declaration:

Reviewer: ```suggestion
The default task factory allows simplifying action nodes declaration with the following syntax sugar:
```

```hocon
action = reference-to-action
# onTransitions { }
```

> A nice syntax sugar!

Reviewer: ```suggestion
```

[logger](https://github.com/Knotx/knotx-fragments/blob/feature/%2347-action-log-structure/handler/api/src/main/java/io/knotx/fragments/handler/api/actionlog/ActionLogBuilder.java) 
implementation that hides syntax complexity.

> #### Action node factory

Reviewer: ```suggestion
#### Subtasks node factory
```

> It is implemented by the [SubtasksNodeFactory](https://github.com/Knotx/knotx-fragments/blob/master/handler/core/src/main/java/io/knotx/fragments/task/factory/node/subtasks/SubtasksNodeFactory.java) 

Reviewer: ```suggestion
A subtask is nothing else than a subgraph defined inside the task.
Creating subtasks is implemented in the [SubtasksNodeFactory](https://github.com/Knotx/knotx-fragments/blob/master/handler/core/src/main/java/io/knotx/fragments/task/factory/node/subtasks/SubtasksNodeFactory.java) 
```
